src/UrqlClient.re

```diff
@@ -2,14 +2,16 @@ open UrqlTypes;
 type t;
 
 /* Helpers for supporting polymorphic fetchOptions. */
-type fetchOptions =
-  | FetchOpts(Fetch.requestInit)
-  | FetchFn(unit => Fetch.requestInit);
+type fetchOptions('a) =
+  | FetchOpts(Fetch.requestInit): fetchOptions(Fetch.requestInit)
+  | FetchFn(unit => Fetch.requestInit)
+    : fetchOptions(unit => Fetch.requestInit);
 
-let unwrapFetchOptions = fetchOptions =>
+let unwrapFetchOptions = (type a, fetchOptions: option(fetchOptions(a))): a =>
   switch (fetchOptions) {
-  | FetchOpts(opts) => opts
-  | FetchFn(fn) => fn()
+  | Some(FetchOpts(opts)) => opts
+  | Some(FetchFn(fn)) => fn
+  | None => Obj.magic(Fetch.RequestInit.make())
```

Comment:
Is there potential we could return an `option` type here to prevent the need for `Obj.magic`, i.e.:

```reason
let unwrapFetchOptions = (type a, fetchOptions: option(fetchOptions(a))): option(a) =>
   switch (fetchOptions) {
   | Some(FetchOpts(opts)) => Some(opts)
   | Some(FetchFn(fn)) => Some(fn)
   | None => None
   }
```

IIRC, BuckleScript will pass along the value wrapped by `Some` to the JS side, and in the `None` case will pass `undefined`, which should be totally fine for the `urql` side. I don't mind `Obj.magic` too much, so also ok with this implementation if you think it's better 👍

Reply:
Great suggestion, I made this change, now we're 100% safe :^)